R/build_coxph.R

```diff
@@ -62,6 +62,190 @@ build_coxph <- function(data, formula, max_categories = NULL, min_group_size = N
+  # ...
+  #     1    NA     H1    1.000
+  #     1    NA     H2    0.999
+  # ...
+  #     1    NA     L1    0.984
+  #     1    NA     L2    0.981
+  ret <- ret %>% tidyr::separate(period, sep=1, into=c("type","period"), remove=TRUE) 
+  ret <- ret %>% dplyr::mutate(period = as.numeric(period))
+  # Format of ret looks like this:
+  #     trt   age period survival
+  #   <dbl> <dbl>  <dbl>    <dbl>
+  # 1     1    NA      1    0.997
+  # 2     1    NA      2    0.995
+  # 3     1    NA      3    0.984
+  # 4     1    NA      4    0.981
+  ret <- ret %>% tidyr::pivot_wider(names_from="type", values_from="survival")
+  # Format of ret looks like this:
+  #  `age` `sex` period     V     H     L
+  # <dbl> <fct>  <dbl>  <dbl> <dbl> <dbl>
+  #    39 <NA>       1  0.996     1 0.988
+  #    39 <NA>       2  0.992     1 0.979
+  #    39 <NA>       3  0.988     1 0.972
+  #    39 <NA>       4  0.984     1 0.964
+  ret <- ret %>% dplyr::rename(survival=V, conf.high=H, conf.low=L)
+
+  chart_type_map <- c()
+  for(col in colnames(ret)) {
+    chart_type_map <- c(chart_type_map, is.numeric(ret[[col]]))
+  }
+  chart_type_map <- ifelse(chart_type_map, "line", "scatter")
+  names(chart_type_map) <- colnames(ret)
+
+  ret <- ret %>% tidyr::pivot_longer(c(-period, -survival, -conf.high, -conf.low) ,names_to = 'variable', values_to = 'value', values_ptype = list(value=character()), values_drop_na=TRUE)
+  # Format of ret looks like this:
+  #   period survival variable value
+  #   <chr>     <dbl> <chr>    <dbl>
+  # 1 1         0.997 trt          1
+  # 2 2         0.995 trt          1
+  # 3 3         0.984 trt          1
+  # 4 4         0.981 trt          1
+  # 5 5         0.963 trt          1
+  # 6 6         0.938 trt          1
+  # 7 7         0.931 trt          1
+  # 8 8         0.920 trt          1
+  # 9 9         0.902 trt          1
+  #10 10        0.896 trt          1
+  ret <- ret %>%  dplyr::mutate(period = (!!times)[period]) # Map back period from index to actual time.
+  ret <- ret %>%  dplyr::mutate(chart_type = chart_type_map[variable])
+  ret
+}
+
+# Efron's partial log likelihood approximation.
+# We use this as the cost function for permutation importance of Cox regression.
+# References:
+# http://www.math.ucsd.edu/~rxu/math284/slect5.pdf
+# https://bydmitry.github.io/efron-tensorflow.html
+# https://en.wikipedia.org/wiki/Proportional_hazards_model
+# 
+# lp - Linear predictor
+calc_efron_log_likelihood <- function(lp, time, status) {
+  if (is.data.frame(time)) { # Since mmpf::permutationImportance passes down time as tibble, convert it to vector. TODO: Do this at more appropriate place.
+    time <- time[[1]]
+  }
+  tmp_df <- tibble::tibble(time=time,
+                           status=as.numeric(status), # Handle logical status. TODO: Handle it better.
+                           lp=lp,
+                           theta=exp(lp))
+  tmp_df <- tmp_df %>% dplyr::group_by(time) %>%
+    dplyr::summarize(sum_lp_event = sum(lp*status),
```

Comment:
Make sure those values have no NA, or NA is one of expected values as a result of sum().

Reply:
Added na.rm = TRUE just in case.